s3box_review/s3box.go

```diff
@@ -0,0 +1,243 @@
+	s3Handler *s3.S3
+
+	// bufferSize is the maximum size of data we're willing to buffer before creating an s3 file
+	bufferSize int
+
+	// bufferedData is the data currently buffered in the box. Calling Dump ships this data into s3
+	bufferedData []byte
+
+	// timestamp tracks the time a box was created or reset
+	timestamp time.Time
+
+	// fileNumber indicates the number of s3 files which have currently been created
+	fileNumber int
+
+	// fileLocations stores the s3 files already created
+	fileLocations []string
+
+	// isSealed indicates whether writes are currently allows to the buffer
+	isSealed bool
+}
+
+// NewS3BoxOptions is the expected input for creating a new S3Box.
+// Currently only an S3Bucket is required. If AWS vars aren't explicitly provided, they'll
+// be pulled from your environment.
+type NewS3BoxOptions struct {
+	// S3Bucket is the destination s3 bucket.
+	// This is required.
+	S3Bucket string
+
+	// AWSKey is the AWS ACCESS KEY ID.
+	// By default grabs from your environment.
+	AWSKey string
+
+	// AWSPassword is the AWS SECRET ACCESS KEY.
+	// By default grabs from your environment.
+	AWSPassword string
+
+	// AWSToken is the AWS SESSION TOKEN.
+	// By default grabs from your environment.
+	AWSToken string
+
+	// BufferSize is the maximum size of data we buffer internally
+	// before creating an s3 file.
+	// This is optional and defaults to 10MB.
+	BufferSize int
+}
+
+// NewS3Box creates a new S3Box given the input options.
+// Errors occur if there's an invalid input or if there's difficulty setting up an s3 connection.
+func NewS3Box(options NewS3BoxOptions) (*S3Box, error) {
+	// Check for required inputs and a valid destination config
+	if options.S3Bucket == "" {
+		return nil, ErrS3BucketRequired
+	}
+
+	bufferSize := DefaultBufferSize
+	if options.BufferSize > 0 {
+		bufferSize = options.BufferSize
+	}
+
+	// Setup s3 handler and aws configuration. If no creds are explicitly provided, they'll be grabbed from the environment.
+	region, err := getRegionForBucket(options.S3Bucket)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get AWS region for bucket %s: (%s)", options.S3Bucket, err)
+	}
+
+	// If AWS creds were provided use those, otherwise grab them from your environment
+	var awsCreds *credentials.Credentials
+	if options.AWSKey == "" && options.AWSPassword == "" && options.AWSToken == "" {
+		awsCreds = credentials.NewEnvCredentials()
+	} else {
+		awsCreds = credentials.NewStaticCredentials(options.AWSKey, options.AWSPassword, options.AWSToken)
+	}
+	awsConfig := aws.NewConfig().WithRegion(region).WithS3ForcePathStyle(true).WithCredentials(awsCreds)
+	awsSession := session.New()
+
+	return &S3Box{
+		s3Bucket:   options.S3Bucket,
+		timestamp:  time.Now(),
+		s3Handler:  s3.New(awsSession, awsConfig),
+		bufferSize: bufferSize,
+	}, nil
+}
+
+// NextBox gives you a new box, forgetting everything about previously packaged data
+func (sb *S3Box) NextBox() {
```

Comment:
I'm going to get rid of it until a clear use case is needed. In order to go along with this suggestion we'd be returning a new box for the user to use, rather than clearing out the existing one.